crates/ruma-state-res/tests/it/resolve.rs

```diff
@@ -0,0 +1,341 @@
+            .then(self.state_key.cmp(&other.state_key))
+            .then(self.event_id.cmp(&other.event_id))
+    }
+}
+
+impl PartialOrd for ResolvedState {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Information to be captured in snapshot assertions
+struct Snapshots {
+    /// The resolved state of the room.
+    resolved_state: BTreeSet<ResolvedState>,
+}
+
+/// Test a list of JSON files containing a list of PDUs and return the results.
+///
+/// State resolution is run both atomically for all PDUs and in batches of PDUs by file.
+fn run(paths: &[&str]) -> Snapshots {
+    // Run `cargo test -- --show-output` to view traces, set `RUST_LOG` to control filtering.
+    let _subscriber = tracing_subscriber::fmt()
+        .with_env_filter(EnvFilter::from_default_env())
+        .with_test_writer()
+        .finish()
+        .set_default();
+
+    // Wrapping a fallible function, as opposed to returning `Result` to the test harness, allows us
+    // to more easily control the output format of errors.
+    match try_run(paths) {
+        Ok(x) => x,
+        Err(e) => panic!("Error: {e}"),
+    }
+}
+
+/// Fallible implementation of [`run`], see that function's docs for details.
+fn try_run(paths: &[&str]) -> Result<Snapshots, Box<dyn Error>> {
+    let fixtures_path = Path::new("tests/it/fixtures");
+
+    let pdu_batches = paths
+        .iter()
+        .map(|x| from_json_str(&fs::read_to_string(fixtures_path.join(x))?).map_err(Into::into))
+        .collect::<Result<Vec<Vec<Pdu>>, Box<dyn Error>>>()?;
+
+    let room_version = {
+        let first_pdu = pdu_batches
+            .first()
+            .ok_or("there should be at least one file of PDUs")?
+            .first()
+            .ok_or("there should be at least one PDU in the first file")?;
+
+        if first_pdu.kind != TimelineEventType::RoomCreate {
+            return Err("the first PDU in the first file should be an m.room.create event".into());
+        }
+
+        from_json_str::<ExtractRoomVersion>(first_pdu.content.get())
+            .map_err(|e| format!("the m.room.create PDU's content is invalid: {e}"))?
+            .room_version
+    };
+
+    // Resolve PDUs in batches by file
+    let mut pdus_by_id = HashMap::new();
+    let mut batched_resolved_state = None;
+    for pdus in &pdu_batches {
+        batched_resolved_state =
+            Some(resolve_batch(&room_version, pdus, &mut pdus_by_id, &mut batched_resolved_state)?);
+    }
+    let batched_resolved_state =
+        batched_resolved_state.ok_or("state resolution should have run at least once")?;
+
+    // Resolve all PDUs in a single step
+    let atomic_resolved_state = resolve_batch(
+        &room_version,
+        pdu_batches.iter().flat_map(|x| x.iter()),
+        &mut HashMap::new(),
+        &mut None,
+    )?;
+
+    // Reshape the data a bit to make the snapshots easier to compare.
+    let reshape = |x: StateMap<_>| {
```

Comment:
I don't know if that's a good idea because it means we'd need to construct 2 entire `Snapshots` just to diff the `BTreeSet<ResolvedState>`s. Again keep in mind that `Snapshots` is intended for future expansion to contain other fields that will likely require more computation to generate. I think it's unlikely that, if `resolved_state`s are equal, other fields of `Snapshots` could be unequal, so I think it would be wasteful to generate and compare entire `Snapshots`.

Reply:
Okay then.